### PR TITLE
CF-609E C++ Solution

### PR DESCRIPTION
**Pull request by envyaims**
_If any of the below doesn't apply to this Pull Request, mark the checkbox as done._

- [x] I have tested my code.
- [x] I have added my solution according to the steps [here](https://usaco.guide/general/adding-solution#steps).
- [x] I have followed the code conventions mentioned [here](https://usaco.guide/general/adding-solution/#code-conventions), which includes the following: 
  - I understand that if it is clear that I have not attempted to follow these guidelines (ex. if I have not used tabs to indent), my PR will be closed.
  - If changes are requested, I will re-request the review after making them.

the official editorial doesn't provide a sample code

### CODE REVIEW

File: solutions/cf-609E.mdx [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+const int MAXL = 20; // approximately maximum log
+
+// for euler-tour
+array<int, MAXN> enter_time, exit_time, depth;
+
+// for binary lifting
+array<array<int, MAXL>, MAXN> up, max_up;
+// up[i][j] stores the 2^jth ancestor of vertex i
+// max_up[i][j] stores the maximum weight of all the edges
+// from i to its 2^jth ancestor
```
Comment (SansPapyrus683) [inline solutions/cf-609E.mdx:36]: 1. put these above the variable declarations
2. separate lines for var declarations

File: solutions/cf-609E.mdx [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+<LanguageSection>
+
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+using ll = long long;
+
+const int MAXN = 2e5 + 1;
```
Comment (SansPapyrus683) [inline solutions/cf-609E.mdx:26]: MAXN is really 2e5, so either 0-index or use MAXN + 1 where needed instead

File: solutions/cf-609E.mdx [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+ll kruskal(vector<array<int,4>>& edges){
+	
+	// sort by non-decreasing weight
+	sort(edges.begin(), edges.end(), [&](auto a, auto b){
+		return a[3] < b[3];
+	});
+	
+	// implementation of basic Disjoint Set Union
+	vector<int> parent(n+1), sz(n+1, 1);
+	iota(parent.begin(), parent.end(), 0);
```
Comment (SansPapyrus683) [inline solutions/cf-609E.mdx:54]: iota? if this does what i think it does, c++ already fills the vector with 0

File: solutions/cf-609E.mdx [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+			x = up[x][i];
+		}
+	}
+	return res;
+}
+
+int main(){
+	ios::sync_with_stdio(false);
+	cin.tie(0);
+
```
Comment (SansPapyrus683) [inline solutions/cf-609E.mdx:139]: remove if this runs w/o

File: solutions/cf-609E.mdx [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+	vector<ll> ans(m);
+	for(array<int, 4> edge : edges){
+		int index = edge[0], u = edge[1], v = edge[2], w = edge[3];
+		int least_common_ancestor = lca(u, v);
+		int max_u = getmaxup(u, depth[u] - depth[least_common_ancestor]);
+		int max_v = getmaxup(v, depth[v] - depth[least_common_ancestor]);
+		ans[index] = total_cost - max(max_u, max_v) + w;
+	}
+	
+	for(int i = 0; i < m; i++){
```
Comment (SansPapyrus683) [inline solutions/cf-609E.mdx:171]: spaces for control flow statements

File: solutions/cf-609E.mdx [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+	for(int k = 1; k < MAXL; k++){
+		for(int i = 1; i <= n; i++){
+			up[i][k] = up[up[i][k - 1]][k - 1];
+			// take max of weights from left and right
+			max_up[i][k] = max(max_up[i][k - 1], max_up[up[i][k - 1]][k - 1]);
+		}
+	}
+	
+	vector<ll> ans(m);
+	for(array<int, 4> edge : edges){
```
Comment (SansPapyrus683) [inline solutions/cf-609E.mdx:163]: maybe use a struct for the edges

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]

File: solutions/cf-609E.mdx (hunk 3 above)
Comment (envyaims) [reply solutions/cf-609E.mdx:54]: search up the cpp docs

File: solutions/cf-609E.mdx (hunk 6 above)
Comment (envyaims) [reply solutions/cf-609E.mdx:163]: i explained how the array works in the main function. is it necessary?

File: solutions/cf-609E.mdx (hunk 2 above)
Comment (envyaims) [reply solutions/cf-609E.mdx:26]: well its the MAXN if one indexed, I can add a comment but i think using MAXN + 1 everywhere will be more confusing?

Comment (SansPapyrus683) [reply solutions/cf-609E.mdx:26]: or you could just, yknow, 0-index

File: solutions/cf-609E.mdx (hunk 6 above)
Comment (SansPapyrus683) [reply solutions/cf-609E.mdx:163]: or just use a struct with meaningful names

File: solutions/cf-609E.mdx (hunk 2 above)
Comment (envyaims) [reply solutions/cf-609E.mdx:26]: but ill be adding one to more things and things become uncool

Comment (SansPapyrus683) [reply solutions/cf-609E.mdx:26]: just sub 1 from the node indices?

File: solutions/cf-609E.mdx (hunk 5 above)
Comment (envyaims) [reply solutions/cf-609E.mdx:171]: example?

Comment (SansPapyrus683) [reply solutions/cf-609E.mdx:171]:
```suggestion
	for (int i = 0; i < m; i++) {
```

File: solutions/cf-609E.mdx [hunk 7]
```diff
@@ -23,50 +23,57 @@ the path from vertex $u$ to $v$ and add the given weight. This will ensure that
...
 // from i to its 2^jth ancestor
+array<array<int, MAXL>, MAXN> max_up;
 
-map<pair<int, int>, int> weight; // weight of edge {i, j}
+map<pair<int, int>, int> weights; // weight of edge {i, j}
 vector<int> graph[MAXN]; // adjacency list for the original MST
 int timer = 1;
 int n, m;
 
+struct edge {
```
Comment (SansPapyrus683) [inline solutions/cf-609E.mdx:45]: rename to `Edge`

File: solutions/cf-609E.mdx [hunk 8]
```diff
@@ -23,50 +23,57 @@ the path from vertex $u$ to $v$ and add the given weight. This will ensure that
...
 array<int, MAXN> enter_time, exit_time, depth;
 
-// for binary lifting
-array<array<int, MAXL>, MAXN> up, max_up;
+// for binary lifting:
 // up[i][j] stores the 2^jth ancestor of vertex i
+array<array<int, MAXL>, MAXN> up;
+
 // max_up[i][j] stores the maximum weight of all the edges
 // from i to its 2^jth ancestor
```
Comment (SansPapyrus683) [inline solutions/cf-609E.mdx:37]:
```cpp
/*
 * multiline comments are done
 * like this
 */
```

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]
nice!

File: solutions/cf-609E.mdx [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+int timer = 1;
+int n, m;
+
+struct Edge {
+	// bidirectional edge from u to v
+	int index, u, v, weight;
+};
+
+// here we build the MST using kruskals' algorithm
+// it returns the sum of the weights that took to build the MST
```
Comment (SansPapyrus683) [inline solutions/cf-609E.mdx:53]: multiline comment here as well

File: solutions/cf-609E.mdx [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+// up[i][j] stores the 2^jth ancestor of vertex i
+array<array<int, MAXL>, MAXN> up;
+
+/*
+ * max_up[i][j] stores the maximum weight of all the edges
+ * from i to its 2^jth ancestor
+ */ 
+array<array<int, MAXL>, MAXN> max_up;
+
+map<pair<int, int>, int> weights; // weight of edge {i, j}
```
Comment (SansPapyrus683) [inline solutions/cf-609E.mdx:42]: two spaces after code before starting inline comment
```suggestion
map<pair<int, int>, int> weights;  // weight of edge {i, j}
```

File: solutions/cf-609E.mdx [hunk 11]
```diff
@@ -0,0 +1,183 @@
...
+using ll = long long;
+
+const int MAXN = 2e5;
+const int MAXL = 20; // approximately maximum log
+
+// for euler-tour
+array<int, MAXN> enter_time, exit_time, depth;
+
+// for binary lifting:
+// up[i][j] stores the 2^jth ancestor of vertex i
```
Comment (SansPapyrus683) [inline solutions/cf-609E.mdx:33]: multiline comment or just
```suggestion
// up[i][j] stores the 2^jth ancestor of vertex i (for binary lifting)
```

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]

File: solutions/cf-609E.mdx (hunk 9 above)
Comment (envyaims) [reply solutions/cf-609E.mdx:53]: its two different statements though

File: solutions/cf-609E.mdx [hunk 12]
```diff
@@ -0,0 +1,182 @@
...
+	iota(parent.begin(), parent.end(), 0);
+	
+	function<int(int)> get = [&](int x) {
+		if (parent[x] != x) {
+			parent[x] = get(parent[x]);
+		}
+		return parent[x];
+	};
+	
+	// returns true if both vertexes are not already united
```
Comment (maggie-j-liu) [inline solutions/cf-609E.mdx:71]:
```suggestion
	// returns true if the vertices are not already united
```

File: solutions/cf-609E.mdx [hunk 13]
```diff
@@ -0,0 +1,182 @@
...
+
+bool is_ancestor(int x, int y) {
+	return enter_time[x] <= enter_time[y] && exit_time[y] <= exit_time[x];
+}
+
+int lca(int x, int y) {
+	if (is_ancestor(x, y)) {
+		return x;
+	}
+	for (int i = MAXL-1; i >= 0; i--) {
```
Comment (maggie-j-liu) [inline solutions/cf-609E.mdx:123]:
```suggestion
	for (int i = MAXL - 1; i >= 0; i--) {
```

File: solutions/cf-609E.mdx [hunk 14]
```diff
@@ -0,0 +1,182 @@
...
+		if (!is_ancestor(up[x][i], y)) {
+			x = up[x][i];
+		}
+	}
+	return up[x][0];
+}
+
+// get maximum weight of all the edges from x to its kth ancestor
+int getmaxup(int x, int k) {
```
Comment (maggie-j-liu) [inline solutions/cf-609E.mdx:132]: (optional) rename function to `get_max_up` to match the other functions

**Review by maggie-j-liu — CHANGES_REQUESTED** [review]

File: solutions/cf-609E.mdx [hunk 15]
```diff
@@ -0,0 +1,182 @@
+---
+id: cf-609E
+source: CF
+title: Minimum spanning tree for each edge
+author: Chongtian Ma
+---
+
+[Official Analysis](https://codeforces.com/blog/entry/22187)
```
Comment (ryanchou-dev) [inline solutions/cf-609E.mdx:8]: add lang

File: solutions/cf-609E.mdx [hunk 16]
```diff
@@ -0,0 +1,182 @@
+---
+id: cf-609E
+source: CF
+title: Minimum spanning tree for each edge
```
Comment (ryanchou-dev) [inline solutions/cf-609E.mdx:4]: not the correct title

File: solutions/cf-609E.mdx [hunk 17]
```diff
@@ -0,0 +1,182 @@
...
+author: Chongtian Ma
+---
+
+[Official Analysis](https://codeforces.com/blog/entry/22187)
+
+## Implementation 
+
+As mentioned in the official editorial, the answer for each edge is removing the heaviest edge in the MST in
+the path from vertex $u$ to $v$ and add the given weight. This will ensure that the MST is still connected.
```
Comment (ryanchou-dev) [inline solutions/cf-609E.mdx:13]: if it's already mentioned in the editorial, why mention it here?

at least add some extra intuition/insight

also move out of implementation

File: solutions/cf-609E.mdx [hunk 18]
```diff
@@ -0,0 +1,182 @@
...
+
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+using ll = long long;
+
+const int MAXN = 2e5;
+const int MAXL = 20; // approximately maximum log
```
Comment (ryanchou-dev) [inline solutions/cf-609E.mdx:27]: two spaces for inline comments

File: solutions/cf-609E.mdx [hunk 19]
```diff
@@ -0,0 +1,182 @@
...
+		int u, v, w; 
+		cin >> u >> v >> w;
+		edges[i] = {i, --u, --v, w};
+		weights[{u, v}] = w;
+		weights[{v, u}] = w;
+	}
+	ll total_cost = kruskal(edges);
+	
+	euler_tour(0, 0);
+	// fill binary-lifting arrays
```
Comment (ryanchou-dev) [inline solutions/cf-609E.mdx:157]:
```suggestion
	
	// fill binary-lifting arrays
```

**Review by ryanchou-dev — CHANGES_REQUESTED** [review]

File: solutions/cf-609E.mdx (hunk 19 above)
Comment (SansPapyrus683) [reply solutions/cf-609E.mdx:157]: (optional imo)

File: solutions/cf-609E.mdx (hunk 15 above)
Comment (envyaims) [reply solutions/cf-609E.mdx:8]: the editorial doesnt provide a sample code...

File: solutions/cf-609E.mdx (hunk 17 above)
Comment (envyaims) [reply solutions/cf-609E.mdx:13]: is it better now?

**Review by dongliu0426 — CHANGES_REQUESTED** [review]

File: solutions/cf-609E.mdx (hunk 15 above)
Comment (ryanchou-dev) [reply solutions/cf-609E.mdx:8]: oops ok mb

**Review by ryanchou-dev — APPROVED** [review]

**Review by maggie-j-liu — APPROVED** [review]

**Pull request merged**